Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -358,6 +527,38 @@ bool TextureMapperLayer::shouldBlend() const
     return m_currentOpacity < 1;
 }
 
+bool TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective() const
+{
+    // Check if 3D scene leaf or if has perspective and has children and has 3D transformation
+    if (((!m_state.preserves3D && (m_parent && m_parent->preserves3D())) || m_layerTransforms.localTransform.hasPerspective())
+        && !m_children.isEmpty() && has3DLocalTransform())
+            return true;
+    return false;
+}
+
+bool TextureMapperLayer::has3DLocalTransform() const
+{
+    return !(!m_layerTransforms.localTransform.m13() && !m_layerTransforms.localTransform.m23()
```

Comment:
There are no existing helpers for that? I doubt it.

Reply:
RenderLayer has function called has3DTransform() which returns !m_transformation->isAffine(). Using TransformationMatrix::isAffine() here also probably works but that's not entirely correct because not affine doesn't mean that transformation is 3D. 3D tranformation can also be affine transformation

Reply:
I'd then add a new helper to `TransformationMatrix`  directly.

There is already:

```
    enum class Type : uint8_t {
        IdentityOrTranslation,
        Affine, 
        Other
    };
    
    Type type() const
    {
        if (!m13() && !m14() && !m23() && !m24() && !m34() && !m31() && !m32() && m33() == 1 && m44() == 1) {
            if (!m12() && !m21() && m11() == 1 && m22() == 1)
                return Type::IdentityOrTranslation;
            if (!m43()) 
                return Type::Affine;
        }
        return Type::Other;
    }
```